microprot/scripts/processing.py

```diff
@@ -1,45 +1,155 @@
+
+    Parameters
+    ----------
+    seqs : str
+        list of skbio Sequence
+    outfile : str
+        file path to output multi-sequence FASTA file
+    """
+    def outseqs():
+        for seq in seqs:
+            yield seq
+    io.write(outseqs(), format='fasta', into=outfile)
+
+
+def read_representatives(represent):
+    """ Read representative protein IDs from file or server
+
+    Parameters
+    ----------
+    represent : str
+        file path to list of representative protein IDs in XML format, or an
+        integer representing the clustering level by which the list will be
+        downloaded from the PDB server
+
+    Return
+    ------
+    list of str
+        protein IDs
+
+    Raises
+    ------
+    ValueError
+        if neither repfile nor cluster is specified
+    """
+    ids = set()
+    xmldoc = None
+    if os.path.isfile(represent):  # read from a local file
+        xmldoc = minidom.parse(represent)
+    elif represent.isdigit():  # download from the PDB server
+        url_str = ('http://www.rcsb.org/pdb/rest/representatives?cluster=%s'
+                   % represent)
+        xml_str = urllib.request.urlopen(url_str).read()
+        xmldoc = minidom.parseString(xml_str)
+    else:
+        raise ValueError('Error: You must specify a local file path or the '
+                         'clustering level.')
+    for x in xmldoc.getElementsByTagName('pdbChain'):
+        l = x.attributes['name'].value.split('.')
+        ids.add('%s_%s' % (l[0].lower(), l[1]))
+    return sorted(list(ids))
+
+
 @click.command()
 @click.option('--infile', '-i', required=True,
-              type=click.Path(resolve_path=True, readable=True, exists=True,
-                              file_okay=True),
-              help='Input protein sequence file in FASTA format')
-@click.option('--seqidx', '-s', default=0,
-              help='Extract n-th sequence (default: all sequences)')
-def _main(infile, outfile, seqidx):
+              type=click.Path(resolve_path=True, readable=True, exists=True),
+              help='Input protein sequence file in FASTA format.')
+@click.option('--outfile', '-o', required=True,
+              type=click.Path(resolve_path=True, readable=True, exists=False),
+              help='Output protein sequence file in FASTA format.')
+@click.option('--identifiers', '-d', required=False,
+              help='Indexes or IDs of protein sequences to extract. May be a '
+              'comma-separated list, or an external file with one entry per '
+              'line. If not specified, all sequences will be extracted.')
+@click.option('--represent', '-r', required=False,
+              help='Generate representative protein sequences. One may either '
+              'provide a local protein ID list file (XML format), or let the '
+              'program download it from the PDB server by specifying a '
+              'desired clustering level.')
+def _processing(infile, outfile, identifiers, represent):
```

Comment:
instead of writing directly to STDOUT with print, I would always pass a filehandle to the function onto which output could be produced. By default, that should be sys.stdout, but for unittesting you can pass a StringIO, execute the function and compare the value of the StingIO after execution.
Here is an example:

`def build_ncbi_tree(nodes, verbose=False, out=sys.stdout):
    """ Build a TreeNode from a dict of nodes.

    Parameters
    ----------
    nodes : dict(ID: parentID)
        Dictionary containing the whole taxonomy. Key is node taxID, value is
        parent taxID.
    verbose : Boolean
        Print verbose status information while executing. Default = False
    out : file handle
        File handle into verbosity information should be printed.

    Returns
    -------
    A skbio.tree.TreeNode object, holding the whole taxonomy.
    """

    parents = set(nodes.values())
    tips = set(nodes) - parents

    if verbose:
        out.write("build ncbi tree for %i tips: " % len(list(tips)))
    ls = {}
    for c, tip in enumerate(list(tips)):
        if verbose and (int(c % (len(tips) / 100)) == 0):
            out.write(".")
        try:
            ls[tip] = get_lineage(tip, nodes)[:-1]
        except KeyError:
            raise KeyError("Cannot obtain lineage for taxid %s" % tip)
    tree = TreeNode.from_taxonomy(ls.items())
    if verbose:
        out.write(" done.\n")

    return tree`

And the test:
`        out = StringIO()
        tree = build_ncbi_tree(self.taxonomy, verbose=True, out=out)
        self.assertEqual(out.getvalue().strip(),
                         "build ncbi tree for 7 tips: ....... done.")`